File: backend/function.cpp
/*
* Copyright (c) 2009-2010 David Roberts <d@vidr.cc>
*
* Permission is hereby granted, free of charge, to any person obtaining a copy
* of this software and associated documentation files (the "Software"), to deal
* in the Software without restriction, including without limitation the rights
* to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
* copies of the Software, and to permit persons to whom the Software is
* furnished to do so, subject to the following conditions:
*
* The above copyright notice and this permission notice shall be included in
* all copies or substantial portions of the Software.
*
* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
* IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
* AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
* OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
* THE SOFTWARE.
*/

#include "backend.h"

/**
 * Return a unique ID.
 * 
 * @return  a unique ID
 */
static uint64_t getUID() {
    static uint64_t x = 0;
    return ++x;
}

/**
 * Return the call signature of the given function type. An empty string is
 * returned if the function type appears to be non-prototyped.
 * 
 * @param ty  the function type
 * @return    the call signature
 */
std::string JVMWriter::getCallSignature(const FunctionType *ty) {
    if(ty->isVarArg() && ty->getNumParams() == 0)
        // non-prototyped function
        return "";
    std::string sig;
    sig += '(';
    for(unsigned int i = 0, e = ty->getNumParams(); i < e; i++)
        sig += getTypeDescriptor(ty->getParamType(i));
    if(ty->isVarArg()) sig += "I";
    sig += ')';
    sig += getTypeDescriptor(ty->getReturnType());
    return sig;
}

/**
 * Pack the specified operands of the given instruction into memory. The
 * address of the packed values is left on the top of the stack.
 * 
 * @param inst        the given instruction
 * @param minOperand  the lower bound on the operands to pack (inclusive)
 * @param maxOperand  the upper bound on the operands to pack (exclusive)
 */
void JVMWriter::printOperandPack(const Instruction *inst,
                                 unsigned int minOperand,
                                 unsigned int maxOperand) {
    unsigned int size = 0;
    for(unsigned int i = minOperand; i < maxOperand; i++)
        size += targetData->getTypeAllocSize(
            inst->getOperand(i)->getType());

    printSimpleInstruction("bipush", utostr(size));
    printSimpleInstruction("invokestatic",
                           "lljvm/runtime/Memory/allocateStack(I)I");
    printSimpleInstruction("dup");

    for(unsigned int i = minOperand; i < maxOperand; i++) {
        const Value *v = inst->getOperand(i);
        printValueLoad(v);
        printSimpleInstruction("invokestatic",
            "lljvm/runtime/Memory/pack(I"
            + getTypeDescriptor(v->getType()) + ")I");
    }
    printSimpleInstruction("pop");
}

/**
 * Print a call/invoke instruction.
 * 
 * @param functionVal  the function to call
 * @param inst         the instruction
 */
void JVMWriter::printFunctionCall(const Value *functionVal,
                                  const Instruction *inst) {
    unsigned int origin = isa<InvokeInst>(inst) ? 3 : 1;
    if(const Function *f = dyn_cast<Function>(functionVal)) { // direct call
        const FunctionType *ty = f->getFunctionType();
        
        //for(unsigned int i = origin, e = inst->getNumOperands(); i < e; i++)
        //    printValueLoad(inst->getOperand(i));
        
        for(unsigned int i = 0, e = ty->getNumParams(); i < e; i++)
            printValueLoad(inst->getOperand(i + origin));
        if(ty->isVarArg() && inst)
            printOperandPack(inst, ty->getNumParams() + origin,
                                   inst->getNumOperands());
        
        if(externRefs.count(f))
            printSimpleInstruction("invokestatic",
                getValueName(f) + getCallSignature(ty));
        else
            printSimpleInstruction("invokestatic",
                classname + "/" + getValueName(f) + getCallSignature(ty));
        
        if(getValueName(f) == "setjmp") {
            unsigned int varNum = usedRegisters++;
            printSimpleInstruction("istore", utostr(varNum));
            printSimpleInstruction("iconst_0");
            printLabel("setjmp$" + utostr(varNum));
        }
    } else { // indirect call
        printValueLoad(functionVal);
        const FunctionType *ty = cast<FunctionType>(
            cast<PointerType>(functionVal->getType())->getElementType());
        printOperandPack(inst, origin, inst->getNumOperands());
        printSimpleInstruction("invokestatic",
            "lljvm/runtime/Function/invoke_"
            + getTypePostfix(ty->getReturnType()) + "(II)"
            + getTypeDescriptor(ty->getReturnType()));
    }
}

/**
 * Print a call to an intrinsic function.
 * 
 * @param inst  the instruction
 */
void JVMWriter::printIntrinsicCall(const IntrinsicInst *inst) {
    switch(inst->getIntrinsicID()) {
    case Intrinsic::vastart:
    case Intrinsic::vacopy:
    case Intrinsic::vaend:
        printVAIntrinsic(inst); break;
    case Intrinsic::memcpy:
    case Intrinsic::memmove:
    case Intrinsic::memset:
        printMemIntrinsic(cast<MemIntrinsic>(inst)); break;
    case Intrinsic::flt_rounds:
        printSimpleInstruction("iconst_m1"); break;
    case Intrinsic::dbg_declare:
        // ignore debugging intrinsics
        break;
    case Intrinsic::pow:
    case Intrinsic::exp:
    case Intrinsic::log10:
    case Intrinsic::log:
    case Intrinsic::sqrt:
        printMathIntrinsic(inst); break;
    case Intrinsic::bswap:
        printBitIntrinsic(inst); break;
    default:
        errs() << "Intrinsic = " << *inst << '\n';
        llvm_unreachable("Invalid intrinsic function");
    }
}

/**
 * Print a call instruction.
 * 
 * @param inst  the instruction
 */
void JVMWriter::printCallInstruction(const Instruction *inst) {
    if(isa<IntrinsicInst>(inst))
        printIntrinsicCall(cast<IntrinsicInst>(inst));
    else
        printFunctionCall(inst->getOperand(0), inst);
}

/**
 * Print an invoke instruction.
 * 
 * @param inst  the instruction
 */
void JVMWriter::printInvokeInstruction(const InvokeInst *inst) {
    std::string labelname = getUID() + "$invoke";
    printLabel(labelname + "_begin");
    printFunctionCall(inst->getOperand(0), inst);
    if(!inst->getType()->isVoidTy())
        printValueStore(inst); // save return value
    printLabel(labelname + "_end");
    printBranchInstruction(inst->getParent(), inst->getNormalDest());
    printLabel(labelname + "_catch");
    printSimpleInstruction("pop");
    printBranchInstruction(inst->getParent(), inst->getUnwindDest());
    printSimpleInstruction(".catch lljvm/runtime/System$Unwind",
          "from "  + labelname + "_begin "
        + "to "    + labelname + "_end "
        + "using " + labelname + "_catch");
}

/**
 * Allocate a local variable for the given function. Variable initialisation
 * and any applicable debugging information is printed.
 * 
 * @param f     the parent function of the variable
 * @param inst  the instruction assigned to the variable
 */
void JVMWriter::printLocalVariable(const Function &f,
                                   const Instruction *inst) {
    const Type *ty;
    if(isa<AllocaInst>(inst) && !isa<GlobalVariable>(inst))
        // local variable allocation
        ty = PointerType::getUnqual(
                 cast<AllocaInst>(inst)->getAllocatedType());
    else // operation result
        ty = inst->getType();
    // getLocalVarNumber must be called at least once in this method
    unsigned int varNum = getLocalVarNumber(inst);
    if(debug >= 2)
        printSimpleInstruction(".var " + utostr(varNum) + " is "
            + getValueName(inst) + ' ' + getTypeDescriptor(ty)
            + " from begin_method to end_method");
    // initialise variable to avoid class verification errors
    printSimpleInstruction(getTypePrefix(ty, true) + "const_0");
    printSimpleInstruction(getTypePrefix(ty, true) + "store", utostr(varNum));
}

/**
 * Print the body of the given function.
 * 
 * @param f  the function
 */
void JVMWriter::printFunctionBody(const Function &f) {
    for(Function::const_iterator i = f.begin(), e = f.end(); i != e; i++) {
        if(Loop *l = getAnalysis<LoopInfo>().getLoopFor(i)) {
            if(l->getHeader() == i && l->getParentLoop() == 0)
                printLoop(l);
        } else
            printBasicBlock(i);
    }
}

/**
 * Return the local variable number of the given value. Register/s are
 * allocated for the variable if necessary.
 * 
 * @param v  the value
 * @return   the local variable number
 */
unsigned int JVMWriter::getLocalVarNumber(const Value *v) {
    if(!localVars.count(v)) {
        localVars[v] = usedRegisters++;
        if(getBitWidth(v->getType()) == 64)
            usedRegisters++; // 64 bit types occupy 2 registers
    }
    return localVars[v];
}

/**
 * Print the block to catch Jump objects (thrown by longjmp).
 * 
 * @param numJumps  the number of setjmp calls made by the current function
 */
void JVMWriter::printCatchJump(unsigned int numJumps) {
    unsigned int jumpVarNum = usedRegisters++;
    printSimpleInstruction(".catch lljvm/runtime/Jump "
        "from begin_method to catch_jump using catch_jump");
    printLabel("catch_jump");
    printSimpleInstruction("astore", utostr(jumpVarNum));
    printSimpleInstruction("aload", utostr(jumpVarNum));
    printSimpleInstruction("getfield", "lljvm/runtime/Jump/value I");
    for(unsigned int i = usedRegisters-1 - numJumps,
                     e = usedRegisters-1; i < e; i++) {
        if(debug >= 2)
            printSimpleInstruction(".var " + utostr(i) + " is setjmp_id_"
                + utostr(i) + " I from begin_method to end_method");
        printSimpleInstruction("aload", utostr(jumpVarNum));
        printSimpleInstruction("getfield", "lljvm/runtime/Jump/id I");
        printSimpleInstruction("iload", utostr(i));
        printSimpleInstruction("if_icmpeq", "setjmp$" + utostr(i));
    }
    printSimpleInstruction("pop");
    printSimpleInstruction("aload", utostr(jumpVarNum));
    printSimpleInstruction("athrow");
    if(debug >= 2)
        printSimpleInstruction(".var " + utostr(jumpVarNum) + " is jump "
            "Llljvm/runtime/Jump; from begin_method to end_method");
}

/**
 * Print the given function.
 * 
 * @param f  the function
 */
void JVMWriter::printFunction(const Function &f) {
    localVars.clear();
    usedRegisters = 0;
    
    out << '\n';
    out << ".method " << (f.hasLocalLinkage() ? "private " : "public ")
        << "static " << getValueName(&f) << '(';
    for(Function::const_arg_iterator i = f.arg_begin(), e = f.arg_end();
        i != e; i++)
        out << getTypeDescriptor(i->getType());
    if(f.isVarArg())
        out << "I";
    out << ')' << getTypeDescriptor(f.getReturnType()) << '\n';
    
    for(Function::const_arg_iterator i = f.arg_begin(), e = f.arg_end();
        i != e; i++) {
        // getLocalVarNumber must be called at least once in each iteration
        unsigned int varNum = getLocalVarNumber(i);
        if(debug >= 2)
            printSimpleInstruction(".var " + utostr(varNum) + " is "
                + getValueName(i) + ' ' + getTypeDescriptor(i->getType())
                + " from begin_method to end_method");
    }
    if(f.isVarArg()) {
        vaArgNum = usedRegisters++;
        if(debug >= 2)
            printSimpleInstruction(".var " + utostr(vaArgNum)
                + " is varargptr I from begin_method to end_method");
    }
    
    // TODO: better stack depth analysis
    unsigned int stackDepth = 8;
    unsigned int numJumps = 0;
    for(const_inst_iterator i = inst_begin(&f), e = inst_end(&f);
        i != e; i++) {
        if(stackDepth < i->getNumOperands())
            stackDepth = i->getNumOperands();
        if(i->getType() != Type::getVoidTy(f.getContext()))
            printLocalVariable(f, &*i);
        if(const CallInst *inst = dyn_cast<CallInst>(&*i))
            if(!isa<IntrinsicInst>(inst)
            && getValueName(inst->getOperand(0)) == "setjmp")
                numJumps++;
    }
    
    for(unsigned int i = 0; i < numJumps; i++) {
        // initialise jump IDs to prevent class verification errors
        printSimpleInstruction("iconst_0");
        printSimpleInstruction("istore", utostr(usedRegisters + i));
    }
    
    printLabel("begin_method");
    printSimpleInstruction("invokestatic",
                           "lljvm/runtime/Memory/createStackFrame()V");
    printFunctionBody(f);
    if(numJumps) printCatchJump(numJumps);
    printSimpleInstruction(".limit stack", utostr(stackDepth * 2));
    printSimpleInstruction(".limit locals", utostr(usedRegisters));
    printLabel("end_method");
    out << ".end method\n";
}

File: java/src/lljvm/runtime/Function.java
/*
* Copyright (c) 2009 David Roberts <d@vidr.cc>
*
* Permission is hereby granted, free of charge, to any person obtaining a copy
* of this software and associated documentation files (the "Software"), to deal
* in the Software without restriction, including without limitation the rights
* to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
* copies of the Software, and to permit persons to whom the Software is
* furnished to do so, subject to the following conditions:
*
* The above copyright notice and this permission notice shall be included in
* all copies or substantial portions of the Software.
*
* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
* IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
* AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
* OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
* THE SOFTWARE.
*/

package lljvm.runtime;

import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.lang.reflect.Modifier;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

import lljvm.util.ReflectionUtils;

/**
 * Provides function pointers for methods.
 * 
 * @author  David Roberts
 */
public final class Function implements Module {
    /** Set of registered classes */
    private final Set<String> registeredClasses = new HashSet<String>();
    /** Map of function signatures to function pointers */
    private final Map<String, Integer> functionPointers
        = new HashMap<String, Integer>();
    /** Map of function pointers to Target objects */
    private final Map<Integer, Target> functionObjects
        = new HashMap<Integer, Target>();
    
    private Memory memory;
    private Context context;
    /**
     * Prevent this class from being instantiated.
     */
    public Function() {
    }
    @Override
    public void initialize(Context context) {
        this.context = context;
        this.memory = context.getModule(Memory.class);
    }

    @Override
    public void destroy(Context context) {
    }
    //TODO: Have to be able to register instances...
    
    /**
     * If the class specified by the given binary name has not yet been
     * registered, then generate function pointers for all public static
     * methods that it declares.
     * 
     * @param classname  Binary name of the class to register.
     * @throws ClassNotFoundException
     *                   if no class with the given name can be found
     */
    private void registerClass(String classname)
    throws ClassNotFoundException {
        if(registeredClasses.contains(classname))
            return;
        Class<?> cls = Class.forName(classname.replace('/', '.'));
        for(Method method : cls.getDeclaredMethods()) {
            final int addr = memory.allocateData();
            final String sig = ReflectionUtils.getQualifiedSignature(method);
            functionPointers.put(sig, addr);
            functionObjects.put(addr, new Target(method));
            method.setAccessible(true);
        }
        registeredClasses.add(classname);
    }
    
    /**
     * Return a function pointer for the method with the specified signature
     * declared by the specified class.
     * 
     * @param classname        the binary name of the declaring class
     * @param methodSignature  the signature of the method
     * @return                 a function pointer for the specified method
     */
    public int getFunctionPointer(String classname,
                                         String methodSignature) {
        try {
            registerClass(classname);
        } catch(ClassNotFoundException e) {
            throw new RuntimeException(e);
        }
        final String sig = classname + "/" + methodSignature;
        if(!functionPointers.containsKey(sig))
            throw new IllegalArgumentException(
                    "Unable to get function pointer for "+sig);
        return functionPointers.get(sig);
    }
    
    /**
     * Invoke the method pointed to by the given function pointer with the
     * given arguments.
     * 
     * @param f     the function pointer
     * @param args  a pointer to the packed list of arguments
     * @return      the return value of the method
     */
    private Object invoke(int f, int args) {
        final Target target = functionObjects.get(f);
        if(target == null)
            throw new IllegalArgumentException("Invalid function pointer: "+f);
        final Method method = target.getMethod();
        final Object instance = target.getInstance(context);

        final Class<?>[] paramTypes = method.getParameterTypes();
        final Object[] params = memory.unpack(args, paramTypes);
        try {
            return method.invoke(instance, params);
        } catch(IllegalAccessException e) {
            throw new RuntimeException(e);
        } catch(InvocationTargetException e) {
            Throwable cause = e.getCause();
            if(cause instanceof RuntimeException)
                throw (RuntimeException) cause;
            if (cause instanceof java.lang.Error)
                throw (java.lang.Error)cause;
            throw new RuntimeException(cause);
        }
    }
    
    /**
     * Invoke the method pointed to by the given function pointer with the
     * given arguments.
     * 
     * @param f     the function pointer
     * @param args  a pointer to the packed list of arguments
     */
    public void invoke_void(int f, int args) {
        invoke(f, args);
    }
    
    /**
     * Invoke the method pointed to by the given function pointer with the
     * given arguments.
     * 
     * @param f     the function pointer
     * @param args  a pointer to the packed list of arguments
     * @return      the return value of the method
     */
    public boolean invoke_i1(int f, int args) {
        return (Boolean) invoke(f, args);
    }
    
    /**
     * Invoke the method pointed to by the given function pointer with the
     * given arguments.
     * 
     * @param f     the function pointer
     * @param args  a pointer to the packed list of arguments
     * @return      the return value of the method
     */
    public byte invoke_i8(int f, int args) {
        return (Byte) invoke(f, args);
    }
    
    /**
     * Invoke the method pointed to by the given function pointer with the
     * given arguments.
     * 
     * @param f     the function pointer
     * @param args  a pointer to the packed list of arguments
     * @return      the return value of the method
     */
    public short invoke_i16(int f, int args) {
        return (Short) invoke(f, args);
    }
    
    /**
     * Invoke the method pointed to by the given function pointer with the
     * given arguments.
     * 
     * @param f     the function pointer
     * @param args  a pointer to the packed list of arguments
     * @return      the return value of the method
     */
    public int invoke_i32(int f, int args) {
        return (Integer) invoke(f, args);
    }
    
    /**
     * Invoke the method pointed to by the given function pointer with the
     * given arguments.
     * 
     * @param f     the function pointer
     * @param args  a pointer to the packed list of arguments
     * @return      the return value of the method
     */
    public long invoke_i64(int f, int args) {
        return (Long) invoke(f, args);
    }
    
    /**
     * Invoke the method pointed to by the given function pointer with the
     * given arguments.
     * 
     * @param f     the function pointer
     * @param args  a pointer to the packed list of arguments
     * @return      the return value of the method
     */
    public float invoke_f32(int f, int args) {
        return (Float) invoke(f, args);
    }
    
    /**
     * Invoke the method pointed to by the given function pointer with the
     * given arguments.
     * 
     * @param f     the function pointer
     * @param args  a pointer to the packed list of arguments
     * @return      the return value of the method
     */
    public double invoke_f64(int f, int args) {
        return (Double) invoke(f, args);
    }
    
    private static final class Target {
        private final Method method;
        private Object instance;
        private boolean init;
        Target(Method method) {
            super();
            this.method = method;
        }
        Object getInstance(Context context) {
            if (!init) {
                if (Modifier.isStatic(method.getModifiers()))
                    instance = null;
                else {
                    //TODO - Think about inheritance
                    instance = context.getModule(method.getDeclaringClass());
                }
                init = true;
            }
            return instance;
        }
        Method getMethod() {
            return method;
        }        
    }
}
